numba/tests/npyufunc/test_gufunc.py

```diff
@@ -68,6 +68,72 @@ class TestGUFuncParallel(TestGUFunc):
     target = 'parallel'
 
 
+class TestDynamicGUFunc(TestCase):
+    target = 'cpu'
+
+    def test_dynamic_matmul(self):
+
+        def check_matmul_gufunc(gufunc, A, B, C):
+            gufunc(A, B, C)
+            Gold = ut.matrix_multiply(A, B)
+            np.testing.assert_allclose(C, Gold, rtol=1e-5, atol=1e-8)
+
+        gufunc = GUVectorize(matmulcore, '(m,n),(n,p)->(m,p)',
+                             target=self.target)
+
+        matrix_ct = 10
+        A = np.arange(matrix_ct * 2 * 4, dtype=np.float32).reshape(matrix_ct, 2, 4)
```

Comment:
If I change the order of the tests (run the integral type first), then one ensures that two versions will be generated, right?

Reply:
Looks like it is not casting because of the unsafe cast from int64 to float32. The `TypeError` is:

```
ufunc 'matmulcore' not supported for the input types, and the inputs could not be safely coerced to any supported types according to the casting rule ''safe''
```

If the gufunc starts with a float64 outputs and then with int64 output, the `TypeError` will be:

```
ufunc 'matmulcore' output (typecode 'd') could not be coerced to provided output parameter (typecode 'l') according to the casting rule ''same_kind''
```